File: ruby/234_palindrome_linked_list.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val, nextNode=nil)
        @val = val
        @next = nextNode
    end
end

# @param {ListNode} head
# @return {Boolean}

# 判断链表中val是否为回文
# 首先利用快慢指针找到另一边的head，将其反转记做 `reverse`，
# 然后head 和 reverse进行比较val即可

# Input: 1->2->2->1
# Output: true

def reverse_list(head)
	left, right = nil, nil
	while head
		left = head
		head = head.next
		left.next = right
		right = left
	end

	right
end

# 判断链表是否为回文，判断回文使用头尾指针，不过因为是链表，我们可以使用判断两边是否相等来判断
# 首先找到链表中间节点，将其倒置，然后将其与head一一对比即可
# 1-2-2-1 	=> middle=2 => 1-2-nil, 1-2-nil
# 1-1-3-1-1	=> middle=3 => 1-1-3-nil, 1-1-3-nil


def is_palindrome(head)
	return true if head.nil? || head.next.nil?

	slow, fast = head, head
	while fast && fast.next
		slow = slow.next
		fast = fast.next.next
	end

	reverse = reverse_list(slow)

	puts show_list_node_val(reverse)
	puts show_list_node_val(head)

	while reverse
		return false if head.val != reverse.val
		head = head.next
		reverse = reverse.next
	end

	return true
end

def show_list_node_val(node)
	str = ""
	while node
		str += node.val.to_s
		node = node.next
		str += '->' if node
	end

	str
end

l1 = ListNode.new(1, ListNode.new(1, ListNode.new(3, ListNode.new(1, ListNode.new(1,nil)))))

puts is_palindrome(l1)

File: cpp/234_palindrome_linked_list.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 给定链表判断是否是回文；题目需要O(n)复杂度，O(1)空间；
// 1> 使用栈，遍历两次看是否一致；
// 2> 找到中间节点，一段反转，然后比较是否一致；

// 1 2 1
// 1 2 2 1
//
//
// 1-2-3
// 1-NULL
// 2-1-null
// 3-2-1-null
class Solution {
public:
	ListNode *reverseList(ListNode *head) {
		ListNode *root = NULL;
		while (head) {
			auto nextNode = head->next;
			head->next = root;
			root = head;
			head = nextNode;
		}
		return root;
	}

    bool isPalindrome(ListNode* head) {
    	if (!head) { return true; }
		auto slow = head, fast = head;
		while (fast->next) {
			if (fast->next->next) {
				fast = fast->next->next;
			} else {
				fast = fast->next;
			}
			slow = slow->next;
		}
		auto rHead = reverseList(slow);

		while (head && rHead) {
			if (head->val != rHead->val) { return false; }
			head = head->next;
			rHead = rHead->next;
		}
		return true;
    }
};
